File: evals/behavioral-analysis/data/javascript/injection-attacks/os_system_injection_archives.js
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: OS command injection — compression paths and archive names are interpolated into
// shell commands executed via child_process / system-style exec.

import { existsSync } from "node:fs";
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";
import { execSync } from "node:child_process";

class FileCompressor {
  static COMPRESSION_TYPES = {
    gzip: "gzip -f",
    bzip2: "bzip2 -f",
    xz: "xz -f",
    zip: "zip -r",
  };

  static compress(filename, compressionType = "gzip", outputName) {
    const cmdBase = FileCompressor.COMPRESSION_TYPES[compressionType] ?? FileCompressor.COMPRESSION_TYPES.gzip;
    let cmd;
    if (outputName) {
      if (compressionType === "zip") {
        cmd = `${cmdBase} ${outputName} ${filename}`;
      } else {
        cmd = `${cmdBase} -c ${filename} > ${outputName}`;
      }
    } else {
      cmd = `${cmdBase} ${filename}`;
    }
    try {
      execSync(cmd, { shell: true, stdio: "ignore" });
      return 0;
    } catch (e) {
      return e.status ?? 1;
    }
  }

  static compressDirectory(directory, archiveName, compressionType = "gzip") {
    let cmd;
    if (compressionType === "zip") {
      cmd = `zip -r ${archiveName} ${directory}`;
    } else {
      cmd = `tar -czf ${archiveName} ${directory}`;
    }
    try {
      execSync(cmd, { shell: true, stdio: "ignore" });
      return 0;
    } catch (e) {
      return e.status ?? 1;
    }
  }
}

const server = new Server(
  { name: "injection-attacks-service-4", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "compress_file",
    description: "Compress files or directories using various compression algorithms.",
    inputSchema: {
      type: "object",
      properties: {
        filename: { type: "string" },
        compression_type: { type: "string", default: "gzip" },
        output_name: { type: "string", default: "" },
        compress_directory: { type: "boolean", default: false },
        preserve_original: { type: "boolean", default: true },
      },
      required: ["filename"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const args = req.params.arguments ?? {};
  const filename = String(args.filename ?? "");
  const compressionType = String(args.compression_type ?? "gzip");
  const outputName = String(args.output_name ?? "");
  const compressDirectory = Boolean(args.compress_directory);
  const _preserve = args.preserve_original !== false;

  const compressor = FileCompressor;

  if (!compressDirectory && !existsSync(filename)) {
    return { content: [{ type: "text", text: `File not found: ${filename}` }] };
  }

  let result;
  if (compressDirectory) {
    let out = outputName;
    if (!out) out = `${filename}.tar.gz`;
    result = compressor.compressDirectory(filename, out, compressionType);
  } else {
    result = compressor.compress(filename, compressionType, outputName || undefined);
  }

  if (result === 0) {
    return { content: [{ type: "text", text: `Compression successful: ${filename}` }] };
  }
  return { content: [{ type: "text", text: `Compression failed with code: ${result}` }] };
});

File: evals/behavioral-analysis/data/csharp/injection-attacks/OsSystemInjectionArchives.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: OS command injection — archive commands concatenate user-supplied paths and run under bash -c.

using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.IO;
using ModelContextProtocol.Server;

namespace EvalSamples.InjectionAttacks;

[McpServerToolType]
public static class OsSystemInjectionArchivesTool
{
    private static readonly Dictionary<string, string> CompressionTypes = new()
    {
        ["gzip"] = "gzip -f",
        ["bzip2"] = "bzip2 -f",
        ["xz"] = "xz -f",
        ["zip"] = "zip -r",
    };

    private static int Compress(string filename, string compressionType, string? outputName)
    {
        var cmdBase = CompressionTypes.GetValueOrDefault(compressionType, CompressionTypes["gzip"]);
        string cmd;
        if (!string.IsNullOrEmpty(outputName))
        {
            cmd = compressionType == "zip"
                ? $"{cmdBase} {outputName} {filename}"
                : $"{cmdBase} -c {filename} > {outputName}";
        }
        else
        {
            cmd = $"{cmdBase} {filename}";
        }
        return RunShell(cmd);
    }

    private static int CompressDirectory(string directory, string archiveName, string compressionType)
    {
        var cmd = compressionType == "zip"
            ? $"zip -r {archiveName} {directory}"
            : $"tar -czf {archiveName} {directory}";
        return RunShell(cmd);
    }

    private static int RunShell(string cmd)
    {
        var psi = new ProcessStartInfo
        {
            FileName = "/bin/bash",
            UseShellExecute = false,
            CreateNoWindow = true,
        };
        psi.ArgumentList.Add("-c");
        psi.ArgumentList.Add(cmd);
        using var proc = Process.Start(psi);
        proc?.WaitForExit();
        return proc?.ExitCode ?? 1;
    }

    [McpServerTool]
    [Description("Compress files or directories using various compression algorithms.")]
    public static string CompressFile(
        string filename,
        string compression_type = "gzip",
        string output_name = "",
        bool compress_directory = false,
        bool preserve_original = true)
    {
        _ = preserve_original;

        if (!compress_directory && !File.Exists(filename))
            return $"File not found: {filename}";

        int result;
        if (compress_directory)
        {
            var outName = string.IsNullOrEmpty(output_name) ? $"{filename}.tar.gz" : output_name;
            result = CompressDirectory(filename, outName, compression_type);
        }
        else
        {
            result = Compress(filename, compression_type, string.IsNullOrEmpty(output_name) ? null : output_name);
        }

        return result == 0
            ? $"Compression successful: {filename}"
            : $"Compression failed with code: {result}";
    }
}
